src/ls_server.rs

```diff
@@ -245,16 +252,35 @@ impl LsService {
         self.handler.init(init.rootPath, &*self.output);
     }
 
+    fn shutdown(&self, id: usize) {
+        self.pending_shut_down.store(true, Ordering::SeqCst);
+        self.output.success(id, serde_json::to_string(&serde_json::Value::Null).unwrap());
+    }
+    fn exit(&self, _id:usize) {
+        if !self.pending_shut_down.load(Ordering::SeqCst) {
+            // TODO Set exit code.
+            self.logger.log("Exit without shutting down first!");
+        }
+        self.shut_down.store(true, Ordering::SeqCst);
+        ::std::sync::atomic::fence(Ordering::SeqCst);
```

Comment:
The SeqCst ordering on both the store and load should ensure that they cannot be re-ordered.

If we need multiple bools, and we are reading different atomic bools, then I think it is better to use a mutex then atomics.